### PR TITLE
feat: display and sign order with partner fee

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

1. Fixed price details accordion displaying (display it only we have a valid trade)
2. Added a total fee amount on the collapsed accordion state (fee + partnerFee)
3. Made the accordion open by default on the confirm screen
4. Changed "Receive (inc. fee)" displayed value and added partner fee into tooltip (sell/buy orders)
5. Removed price info duplication from the accordion
6. Add partner fee into "Expected to receive" and "Minimum receive" values on the confirm screen
7. Added partner fee into appData (it actually was already there, I just fixed the mocking data)

image

image

image

image

image

  # To Test

To avoid confusion, please test through https://github.com/cowprotocol/cowswap/pull/3942
Because there I've deleted non-zero-fee mode, so now we always use fee=zero for swaps.

>In this PR partner-fee is hardcoded as 25%

>https://github.com/cowprotocol/cowswap/pull/3943 - here you can test the changes without mocking partner fee

1. Sell/buy orders
- [ ] "Receive (inc. fee)" should include fee + partnerFee values
- [ ] "From (inc. fee)" should include fee + partnerFee values
2. Confirmation screen
- [ ] "Expected to receive" should include fee + partnerFee values
- [ ] "Minimum receive" should include fee + partnerFee values
- [ ] Sining data should match "Minimum receive"
3. PartnerFee mush be included into appData

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.ts [hunk 1]
```diff
@@ -24,14 +24,14 @@ function inputAmountForSignature(params: AmountForSignatureParams): CurrencyAmou
...
+      return trade.inputAmountWithPartnerFee.multiply(slippageCoeff) // add slippage
     }
   }
 
   if (isSell) {
-    return trade.inputAmountWithFee
+    return trade.inputAmountWithPartnerFee
   } else {
-    return trade.inputAmountWithoutFee.multiply(slippageCoeff) // add slippage
+    return trade.inputAmountWithPartnerFee.multiply(slippageCoeff) // add slippage
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.ts:34]: I'm not sure here

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts [hunk 2]
```diff
@@ -86,8 +106,14 @@ describe('getAmountsForSignature()', () => {
         kind: OrderKind.BUY,
       })
 
-      // inputAmount = (3012) + 5% = 3,162.6
-      expect(result.inputAmount.toFixed()).toEqual('3162.600000')
+      if (partnerFeeBps === 0) {
+        // inputAmount = (3012 + 8) + 5% = 3171.000000
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:110]: I think we should not add fee here

File: apps/cowswap-frontend/src/modules/injectedWidget/hooks/useInjectedWidgetParams.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export function useInjectedWidgetParams(): Partial<CowSwapWidgetParams> {
     ...value,
     // TODO: Remove the default partner fee after testing
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/injectedWidget/hooks/useInjectedWidgetParams.ts:12]: TODO: remove the mock

Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/injectedWidget/hooks/useInjectedWidgetParams.ts:12]: https://github.com/cowprotocol/cowswap/pull/3943

File: apps/cowswap-frontend/src/legacy/state/swap/extension.ts [hunk 4]
```diff
@@ -55,6 +54,8 @@ export function useTradeExactInWithFee({
 
   // external price output as Currency
   const outputAmount = stringToCurrency(quote.price.amount, outputCurrency)
+  const partnerFeeAmount = partnerFee ? outputAmount.multiply(bpsToPercent(partnerFee.bps)) : undefined
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/legacy/state/swap/extension.ts:57]: Can we use `0` for `partnerFeeAmount`, so there's no need for conditional and it's safe to always subtract it from `outputAmount`?

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts [hunk 5]
```diff
@@ -67,17 +81,23 @@ describe('getAmountsForSignature()', () => {
...
       expect(result.inputAmount.toFixed()).toEqual('3004.000000')
 
-      // Fee in WETH = 0.005312084993359893
-      // outputAmount = (2 - 0.005312) - 5% = 1.89495
-      expect(result.outputAmount.toFixed()).toEqual('1.894953519256308100')
+      if (partnerFeeBps === 0) {
+        // outputAmount = 2 - 5% = 1.900000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.900000000000000000')
+      } else {
+        // outputAmount = 2 - 5% - 20% = 1.520000000000000000
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:88]: this is a bit hard to follow, as you are assuming you get a 20% because there's only two tests and one is zero, and the other 20%.

One other way, could be to add this in the `describe.each`

what about something like:

```js
describe.each([
   {partnerFeeBps: 0, expectedOutputAmount: '1.900000000000000000' }, //  explainer here...
   {partnerFeeBps: 2000, expectedOutputAmount: '1.520000000000000000' }, //  explainer here...
])
```

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts [hunk 6]
```diff
@@ -114,16 +142,23 @@ describe('getAmountsForSignature()', () => {
...
 
-      // outputAmount = 2 - 5% = 1.894953
-      expect(result.outputAmount.toFixed()).toEqual('1.894953519256308100')
+      if (partnerFeeBps === 0) {
+        // outputAmount = 2 - 5% = 1.900000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.900000000000000000')
+      } else {
+        // outputAmount = 2 - 5% - 20% = 1.520000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.520000000000000000')
+      }
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:151]: same

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts (hunk 5 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:88]: In the above example, fee is constant. I think is good, as these test would focus on testing variations on the partnerFee for sell orders, but we will need some that focus on variations on the fee too.

Alternativelly, you might want to do a mega-test with all your cases:

describe.each([
   {partnerFeeBps: 0,  fee: 0, slippage: 0, expectedOutputAmount: '1.900000000000000000' }, //  explainer here...
   ... kee some params constants and change the others, to assert the result given some combinations
])

**Review by anxolin — COMMENTED** [review]
This is a great PR!!!!

The labels will change a lot, so feel free to ignore this so we tackle all things together, but I believe we will need to signal this number is an estimation only:

image

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts [hunk 7]
```diff
@@ -55,7 +67,9 @@ describe('getAmountsForSignature()', () => {
      * Output amount is calculated as: (outputAmount - fee) - slippage
      */
     it('Sell order', () => {
-      const trade = getTrade({ tradeType: TradeType.EXACT_INPUT, inputAmount, outputAmount, feeAmount })
+      const trade = getTrade({ tradeType: TradeType.EXACT_INPUT, inputAmount, outputAmount, feeAmount, partnerFeeBps })
+
+      if (!trade) throw new Error('No trade')
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:72]: This test with several conditionals don't look good to me.
Shouldn't we asset the trade exists and fail if it doesn't?

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts [hunk 8]
```diff
@@ -67,17 +81,23 @@ describe('getAmountsForSignature()', () => {
...
-      // Fee in WETH = 0.005312084993359893
-      // outputAmount = (2 - 0.005312) - 5% = 1.89495
-      expect(result.outputAmount.toFixed()).toEqual('1.894953519256308100')
+      if (partnerFeeBps === 0) {
+        // outputAmount = 2 - 5% = 1.900000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.900000000000000000')
+      } else {
+        // outputAmount = 2 - 5% - 20% = 1.520000000000000000
+        expect(result.outputAmount.toFixed()).toEqual('1.520000000000000000')
+      }
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:90]: Here too. These are 2 different tests to me.

File: apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts [hunk 9]
```diff
@@ -28,6 +28,8 @@ export const defaultCurrencyInputPanelProps: CurrencyInputPanelProps & { priceIm
       amountAfterFeesRaw: CurrencyAmount.fromRawAmount(currency, 20 * 10 ** 18),
       feeAmount: '10',
       feeAmountRaw: CurrencyAmount.fromRawAmount(currency, 10 * 10 ** 18),
+      partnerFeeAmount: '25',
+      partnerFeeAmountRaw: CurrencyAmount.fromRawAmount(currency, 25 * 10 ** 18),
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts:32]: Why does this default to non-zero amounts?

**Comment by elena-zh** [comment]
Hey @shoom3301 , this is a great PR!!
(I tested it in #3942) 
However, there are some issues I faced: 
1. There is no fee amount (min) when fees exceed from amount error is displayed (see the comparison with prod)
fee amount
fees example prod
2. There is significant difference in a partner fee calculation for a sell and a buy orders. See details in https://cowservises.slack.com/archives/C0361CDG8GP/p1709127531498219
3. Fee threshold banners should take into account a partner fee (see in https://cowservises.slack.com/archives/C0361CDG8GP/p1709124571171179)
4. I'd add '+gas' into the fee field when an order is onchain (SC wallets, ETH orders) (see https://cowservises.slack.com/archives/C0361CDG8GP/p1709126523835639)
5. when no USD estimation, we decided to have fees in a token value instead of USD estimation 
value in tokens
6. It would be great to enhance values displaying in a mobile view 
image


Thank you!

**Comment by shoom3301** [comment]
@elena-zh thank you!
1. Fixed
2. Fixed
3. Fixed
4. Sorry, I'm not sure how it should looks like. I would keep it for the future, we need Michel here
5. Fixed
6. Fixed only arrow color, I'm not sure about the reset, we also need Michel here

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts (hunk 5 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts]: Refactored the tests, split them into separate cases in WHEN/THEN style

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts (hunk 6 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts:92]: Also refaactored

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts (hunk 7 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts]: This is a part of legacy comes from 'buildTradeExactInWithFee'.
I would make it returning always a trade with no falsy values, but it will require significant refactoring now.
Don't you mind to keep it for the future?

File: apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts (hunk 8 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/swap/helpers/getAmountsForSignature.test.ts]: Fixed

File: apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts (hunk 9 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts:32]: Changed to 0

File: apps/cowswap-frontend/src/legacy/state/swap/extension.ts (hunk 4 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/legacy/state/swap/extension.ts]: I would prefer to keep it more explicit.
If we set 0 here, then it will be difficult to distinguish if the partner fee is not set or set to 0

**Comment by elena-zh** [comment]
Hey @shoom3301 , great!

For the 6th case, I have opened #3953 issue for @fairlighteth to take a look at. 
For the case 4, I found 2 already opened tasks on the board (#1760, #2033)

As for the fixes: 
1. Case 1: could you please add a label to the collapsed dropdown 
image
2. And I can see a negative amount to receive for some trades :( 
receive minus

Could you please fix this?

File: apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts [hunk 10]
```diff
@@ -28,6 +28,8 @@ export const defaultCurrencyInputPanelProps: CurrencyInputPanelProps & { priceIm
       amountAfterFeesRaw: CurrencyAmount.fromRawAmount(currency, 20 * 10 ** 18),
       feeAmount: '10',
       feeAmountRaw: CurrencyAmount.fromRawAmount(currency, 10 * 10 ** 18),
+      partnerFeeAmount: '25',
+      partnerFeeAmountRaw: CurrencyAmount.fromRawAmount(currency, 0),
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts]: This object is a mock for Cosmos.
I've specified values because they are required by the component.

p.s. changed value of partnerFeeAmount to 0

File: apps/cowswap-frontend/src/mocks/tradeStateMock.ts [hunk 11]
```diff
@@ -39,6 +41,8 @@ export const outputCurrencyInfoMock: CurrencyInfo = {
     amountAfterFeesRaw: CurrencyAmount.fromRawAmount(outputCurrency, 20 * 10 ** 18),
     feeAmount: '10',
     feeAmountRaw: CurrencyAmount.fromRawAmount(outputCurrency, 10 * 10 ** 18),
+    partnerFeeAmount: '25',
+    partnerFeeAmountRaw: CurrencyAmount.fromRawAmount(outputCurrency, 0),
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply apps/cowswap-frontend/src/mocks/tradeStateMock.ts:45]: https://github.com/cowprotocol/cowswap/pull/3941#discussion_r1508528292

**Comment by shoom3301** [comment]
@elena-zh thank you! Fixed both

**Comment by elena-zh** [comment]
Hey @shoom3301 , case 1 works great!
As for the case 2, I see calculation mismatch in the tooltip and in the fees dropdown :(
image

**Comment by shoom3301** [comment]
@elena-zh fixed

File: apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts [hunk 12]
```diff
@@ -30,16 +30,16 @@ function useAgnosticPermitDataIfUserHasNoAllowance(): PermitHookData | undefined
 }
 
 export function AppDataHooksUpdater(): null {
-  const { v2Trade } = useDerivedSwapInfo()
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/appData/updater/AppDataHooksUpdater.ts:33]: great!, i hate to see GP trades and gpv2 stuff

**Comment by elena-zh** [comment]
@shoom3301 , still I can face some issues: 
1. Sometimes I can't see estimated fee amount when sell amount is small (see 2 screenshots)
image
image
2. There is an error in percentage calculation for buy orders in the 'Fees exceed' banner
image
According to the formula on Prod, it is calculated based on the From (incl/fee) amount

Could you please fix this?

**Comment by shoom3301** [comment]
@elena-zh thank you!

1. This is not a new issue, it's actually this one: https://cowservices.slack.com/archives/C0361CDG8GP/p1709035605377909. I would like to fix it later, since it requires more analysis and work
2. Fixed

**Comment by elena-zh** [comment]
Hey @shoom3301 , thank you!
However, 
1. The issue here is that est fees field is no longer displayed in some cases, however, it should. And it does not appear when I wait for a while
est fees field
2. I think that still there is something wrong with the percentage calculation. See:
percentage

**Comment by shoom3301** [comment]
@elena-zh thank you!
Fixed both

**Review by elena-zh — APPROVED** [review]
Approved, thank you!

**Review by alfetopito — APPROVED** [review]

File: apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply apps/cowswap-frontend/src/common/pure/CurrencyInputPanel/defaultCurrencyInputProps.ts]: Ah ok, if this is for cosmos it makes sense.
During the review and based on the file name it's not clear this is mock data.
